c_src/device/bcm.c

```diff
@@ -239,67 +236,82 @@ int device_init( const device_opts_t* p_opts, device_info_t* p_info ) {
+  send_reshape(width, height);
 
   // success
   return 0;
 }
 
-int device_close( device_info_t* p_info ) {
+int device_close(device_info_t* p_info)
+{
   return 0;
 }
 
-void device_poll() {}
-
+void device_poll()
+{
+}
 
+void device_begin_render(driver_data_t* p_data)
+{
+  NVGcontext* p_ctx = p_data->p_ctx;
 
-void device_begin_render() {
   glClear(GL_COLOR_BUFFER_BIT);
+
+  nvgBeginFrame(p_ctx, g_device_info.width, g_device_info.height, g_device_info.ratio);
+
+  // set the global transform
+  nvgTransform(p_ctx,
+               p_data->global_tx[0], p_data->global_tx[1],
+               p_data->global_tx[2], p_data->global_tx[3],
+               p_data->global_tx[4], p_data->global_tx[5]);
 }
 
-void device_end_render() {
-  eglSwapBuffers( g_egl_data.display, g_egl_data.surface );
+void device_begin_cursor_render(driver_data_t* p_data)
+{
+  NVGcontext* p_ctx = p_data->p_ctx;
+	nvgTranslate(p_ctx,
+	             p_data->cursor_pos[0], p_data->cursor_pos[1]);
 }
 
+void device_end_render(driver_data_t* p_data)
+{
+  NVGcontext* p_ctx = p_data->p_ctx;
+
+  // End frame and swap front and back buffers
+  //uint64_t time = monotonic_time();
+  nvgEndFrame(p_ctx);
+  //log_info("nvgEndFrame: %" PRId64, monotonic_time() - time);
 
+  //time = monotonic_time();
+  eglSwapBuffers(g_egl_data.display, g_egl_data.surface);
+  //log_info("device_end_render: %" PRId64, monotonic_time() - time);
+}
 
-void device_clear_color( float red, float green, float blue, float alpha ) {
-  glClearColor( red, green, blue, alpha );
+void device_clear_color(float red,
+                        float green,
+                        float blue,
+                        float alpha)
+{
+  glClearColor(red, green, blue, alpha);
 }
 
-char* device_gl_error() {
+char* device_gl_error()
+{
   GLenum err  = glGetError();
   switch (err)
   {
-    case GL_NO_ERROR:
-      return NULL;
-
-    case GL_INVALID_ENUM:
-      return "GL_INVALID_ENUM";
-
-    case GL_INVALID_VALUE:
-      return "GL_INVALID_VALUE";
-
-    case GL_INVALID_OPERATION:
-      return "GL_INVALID_OPERATION";
-
-    case GL_OUT_OF_MEMORY:
-      return "GL_OUT_OF_MEMORY";
-
+  case GL_NO_ERROR: return NULL;
+  case GL_INVALID_ENUM: return "GL_INVALID_ENUM";
+  case GL_INVALID_VALUE: return "GL_INVALID_VALUE";
+  case GL_INVALID_OPERATION: return "GL_INVALID_OPERATION";
+  case GL_OUT_OF_MEMORY: return "GL_OUT_OF_MEMORY";
 #ifdef GL_STACK_UNDERFLOW
-    case GL_STACK_UNDERFLOW:
-      return "GL_STACK_UNDERFLOW";
+  case GL_STACK_UNDERFLOW: return "GL_STACK_UNDERFLOW";
 #endif
-
 #ifdef GL_STACK_OVERFLOW
-    case GL_STACK_OVERFLOW:
-      return "GL_STACK_OVERFLOW";
+  case GL_STACK_OVERFLOW: return "GL_STACK_OVERFLOW";
 #endif
-      
-    case GL_INVALID_FRAMEBUFFER_OPERATION:
-      return "GL_INVALID_FRAMEBUFFER_OPERATION";
-
-    default:
-      return "GL_OTHER";
+  case GL_INVALID_FRAMEBUFFER_OPERATION: return "GL_INVALID_FRAMEBUFFER_OPERATION";
+  default: return "GL_OTHER";
   }
 }
```

Comment:
This routine really wants to live in like a `gl_helpers.c` file.

Reply:
Moved to `c_src/device/gl_helpers.c`